### PR TITLE
[LLP] Allow downstream back-pressuring Indexer

### PR DESCRIPTION
**Pull request by tudor-da**
Performance benchmarks passed for both Indexer and Sandbox-on-X

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala [hunk 1]
```diff
@@ -38,7 +40,8 @@ object BatchingParallelIngestionPipe {
       .async
       .mapAsync(ingestingParallelism)(ingester)
       // Stage 6: Preparing data sequentially for throttled mutations in database (tracking the ledger-end, corresponding sequential event ids and latest-at-the-time configurations)
-      .conflate(tailer)
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala:41]: it seems to me that in this form 2 things can backpressure:
* If downstream (InMemoryState updates and friends) is backpressuring
* If tail ingestion backpressures

I think the second was not intentional, right?

How about we send it to a conflating sink here like:
```
      .alsoTo(
        Flow[DB_BATCH]
          .conflate(tailer)
          .mapAsync(1)(ingestTail)
          .to(Sink.ignore)
      )
```
And then we do batching and/or buffering?

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala [hunk 2]
```diff
@@ -20,8 +20,10 @@ object BatchingParallelIngestionPipe {
       batcher: IN_BATCH => Future[DB_BATCH],
       ingestingParallelism: Int,
       ingester: DB_BATCH => Future[DB_BATCH],
-      tailer: (DB_BATCH, DB_BATCH) => DB_BATCH,
-      ingestTail: DB_BATCH => Future[DB_BATCH],
+      maxTailerBatchSize: Int,
+      tailerSeed: DB_BATCH => Vector[DB_BATCH],
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala:24]: This tailer seed can be fairly trivial if we are zeroing out DB payload already earlier.
How about we zero them out inside the ingester, and then we can remove this injection?

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerSubscription.scala [hunk 3]
```diff
@@ -83,7 +87,10 @@ private[platform] case class ParallelIndexerSubscription[DB_BATCH](
           .buffered(metrics.daml.parallelIndexer.inputBufferLength, maxInputBufferSize)
       )
         .map(batch => batch.offsetsUpdates -> batch.lastSeqEventId)
-        // TODO LLP: Introduce mechanism (buffer) to allow observing downstream backpressure
+        .buffered(
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerSubscription.scala:90]: Is it intentional that we are buffering a batch of batches?
I am a little bit worries about the batchsize * maxTailerBatchSize * maxOutputBatchedBufferSize equation...
How about: 
* we change this to maxOutputBatchSize instead
* we change batching so it is batchWeighted, and weight would be the batch's size, and maxweight would be maxOutputBatchSize
* we remove buffering
* our signal metric for downstream backpressure would be simply the total weight of one weighted batch

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala (hunk 1 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala:41]: > I think the second was not intentional, right?

But it was. The purpose of this PR is to still allow tail ingestion back-pressuring.

Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala:41]: Not sure about that alsoTo solution:
* The order of execution can be inverted if ingestTail is slow leading to in memory state updates ahead of the persisted ledger end
* If ingestTail is very slow, due to conflation, the above problem is amplified

Do we want to deliberately allow in memory state updates front-running persistence ledger end updates?

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerSubscription.scala (hunk 3 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerSubscription.scala:90]: > Is it intentional that we are buffering a batch of batches?

Yes, but I look your idea with batchWeighted better 👍

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala (hunk 1 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala:41]: Discussed offline that it is fine as proposed.

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerSubscription.scala (hunk 3 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerSubscription.scala:93]: Discussed offline that it is fine as proposed.

File: ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala (hunk 2 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala]: Done

**Review by nmarton-da — APPROVED** [review]
LGTM :shipit:

**Pull request merged**